File: python/navigation/global_planning/sampling/ait_star.py
"""AIT* — Adaptively Informed Trees (Strub & Gammell 2020; extended 2022).

What differentiates AIT* from BIT* is the cost-to-go heuristic. BIT* uses the raw
straight-line distance ``h_hat(x) = ||x - goal||``, which ignores obstacles. AIT*
instead computes ``h_hat`` by a **reverse search over the current random geometric
graph (RGG)**: a heuristic that follows the graph's actual connectivity (detours
around a wall) rather than a straight line through it. Crucially it is *adaptive* —
edges found invalid during forward validation are permanently excluded from the
graph the reverse search runs over, so the heuristic self-corrects as obstacles
are discovered (the "lazy reverse search feeds the forward search" loop of Strub &
Gammell 2020, ICRA; extended in IJRR 2022).

Per batch: (1) draw informed samples and append to a persistent point array;
(2) build the RGG neighbour graph, filtering out every edge accumulated in the
persistent ``invalid_edges`` set; (3) reverse Dijkstra from the goal over that
filtered graph produces the adaptive heuristic ``h_hat`` (optimistic — it does not
collision-check, that is the forward search's job); (4) forward A* keyed on
``g[v] + h_hat[v]`` lazily validates each edge it relaxes, adding any invalid edge
to ``invalid_edges`` so it feeds back into every future batch's reverse search.

Implementation simplification (documented, deliberate): the full paper interleaves
a single *incremental* bidirectional LPA*-based search that repairs the reverse
tree and reuses forward ``g`` values across events. Here each batch instead
recomputes the reverse search and the forward ``g``/``parent`` from scratch over
all accumulated samples. This drops the LPA* incrementality (an optimisation of
*how* the searches are updated) while preserving AIT*'s defining behaviour — a
forward search guided by an obstacle-aware reverse heuristic that adapts to
discovered invalid edges — which is what makes it informed and adaptive. Same
guarantee class as BIT*: probabilistically complete, almost-surely asymptotically
optimal, anytime. ``c_best`` (the informed ellipse) and ``invalid_edges`` both
persist across batches and only improve/grow.
"""

from __future__ import annotations

import heapq
import time

import numpy as np

from navigation.core.capabilities import Capability, SamplingSpace
from navigation.core.planner import GlobalPlanner
from navigation.core.trace import TraceRecorder
from navigation.core.types import PlanResult, PlanStats, Point

from ._sampling import informed_sample, path_length, radius_neighbors, rgg_radius

_INF = float("inf")

# Edges are keyed order-independently: an edge is invalid regardless of the
# direction it was traversed when the collision was detected.
Edge = tuple[int, int]


def _edge(i: int, j: int) -> Edge:
    return (i, j) if i < j else (j, i)


def _reverse_search(
    space: SamplingSpace[Point],
    points: list[Point],
    nbr: list[list[int]],
    invalid_edges: set[Edge],
    goal_idx: int,
) -> list[float]:
    """Adaptive cost-to-go heuristic: Dijkstra from the goal over the RGG minus the
    known-invalid edges. Optimistic — it assumes remaining edges are collision-free
    (Strub & Gammell 2020); only the forward search validates, and its findings
    shrink this graph on the next batch."""
    n = len(points)
    h = [_INF] * n
    h[goal_idx] = 0.0
    heap: list[tuple[float, int]] = [(0.0, goal_idx)]
    settled = [False] * n
    while heap:
        d_u, u = heapq.heappop(heap)
        if settled[u]:
            continue
        settled[u] = True
        for w in nbr[u]:
            if _edge(u, w) in invalid_edges:
                continue
            nd = d_u + space.distance(points[u], points[w])
            if nd < h[w]:
                h[w] = nd
                heapq.heappush(heap, (nd, w))
    return h


def _reconstruct(points: list[Point], parent: list[int], goal_idx: int) -> list[Point]:
    path: list[Point] = []
    node = goal_idx
    while node != -1:
        path.append(points[node])
        node = parent[node]
    path.reverse()
    return path


class AITStar(GlobalPlanner[Point, "SamplingSpace[Point]"]):
    @property
    def name(self) -> str:
        return "ait_star"

    def required_capabilities(self) -> set[Capability]:
        return {Capability.SAMPLING_SPACE}

    def plan(
        self,
        space: SamplingSpace[Point],
        start: Point,
        goal: Point,
        recorder: TraceRecorder | None = None,
    ) -> PlanResult[Point]:
        t0 = time.monotonic()
        batch_size = self.params.get_int("batch_size")
        max_batches = self.params.get_int("max_batches")
        gamma = self.params.get_float("gamma")
        rng = np.random.default_rng(self.params.get_int("seed"))

        goal_idx = 1
        points: list[Point] = [start, goal]  # 0 = start (root), 1 = goal
        invalid_edges: set[Edge] = set()  # grows across batches: the adaptive feedback
        c_best = _INF
        expanded = 0
        # Last batch's forward-search result; the final incumbent is read from it.
        g: list[float] = [0.0, _INF]
        parent: list[int] = [-1, -1]

        for _ in range(max_batches):
            # --- 1. grow the RGG (informed once a solution exists) ----------------
            drawn = 0
            for _ in range(batch_size * 40):
                if drawn >= batch_size:
                    break
                q = informed_sample(space, start, goal, c_best, rng)
                if not space.is_state_valid(q):
                    continue
                points.append(q)
                drawn += 1
                if recorder is not None:
                    recorder.sample_drawn(q)

            n = len(points)
            radius = rgg_radius(gamma, n)
            nbr = radius_neighbors(space, points, radius)

            # --- 2+3. reverse search over the filtered graph = adaptive heuristic --
            h_hat = _reverse_search(space, points, nbr, invalid_edges, goal_idx)

            # --- 4. forward A* keyed on g + h_hat, lazily validating each edge -----
            g = [_INF] * n
            g[0] = 0.0
            parent = [-1] * n
            closed = [False] * n
            open_heap: list[tuple[float, int]] = [(h_hat[0], 0)]
            while open_heap:
                key, v = heapq.heappop(open_heap)
                # Lazy deletion: skip entries superseded by a cheaper g[v] (same
                # trick BIT* uses for its vertex queue).
                if closed[v] or key > g[v] + h_hat[v] + 1e-9:
                    continue
                closed[v] = True
                expanded += 1
                for x in nbr[v]:
                    if _edge(v, x) in invalid_edges:
                        continue
                    d = space.distance(points[v], points[x])
                    if not space.is_motion_valid(points[v], points[x]):
                        # Discovered invalid: exclude it from every future batch's
                        # reverse search — this is AIT*'s adaptive feedback loop.
                        invalid_edges.add(_edge(v, x))
                        continue
                    new_g = g[v] + d
                    if new_g < g[x]:
                        first = g[x] == _INF
                        g[x] = new_g
                        parent[x] = v
                        if recorder is not None:
                            # Emit candidate_evaluated only for feasible, improving
                            # edges (not every relaxed neighbour) so the trace stays
                            # renderable by replay.py — matches BIT*'s emit-on-accept
                            # scale rather than exploding on the batch-recomputed graph.
                            recorder.candidate_evaluated(points[x], new_g)
                            if first:
                                recorder.edge_added(points[x], points[v], d)
                            else:
                                recorder.rewire(points[x], points[v])
                        heapq.heappush(open_heap, (new_g + h_hat[x], x))
                        if x == goal_idx and new_g < c_best:
                            c_best = new_g
                            if recorder is not None:
                                recorder.path_found(_reconstruct(points, parent, goal_idx))

        # --- extract final incumbent ---------------------------------------------
        n = len(points)
        success = g[goal_idx] < _INF
        path = _reconstruct(points, parent, goal_idx) if success else []
        cost = path_length(space, path) if success else 0.0
        runtime = time.monotonic() - t0
        if recorder is not None:
            recorder.planning_finished(
                success,
                {
                    "runtime_sec": runtime,
                    "path_cost": cost,
                    "expanded_nodes": float(expanded),
                    "samples": float(n),
                    "tree_size": float(n),
                },
            )
        stats = PlanStats(expanded_nodes=expanded, samples=n, tree_size=n, iterations=expanded)
        return PlanResult(success, path, cost, stats)

File: cpp/src/global_planning/sampling/ait_star.cpp
#include "navigation/global_planning/sampling/ait_star.hpp"

#include <algorithm>
#include <chrono>
#include <cstddef>
#include <functional>
#include <limits>
#include <queue>
#include <set>
#include <utility>
#include <vector>

#include "navigation/global_planning/sampling/sampling_common.hpp"

namespace navigation::global_planning {

namespace {
constexpr double kInf = std::numeric_limits<double>::infinity();

// Edges are keyed order-independently: an edge is invalid regardless of the
// direction it was traversed when the collision was detected.
using Edge = std::pair<int, int>;
Edge make_edge(int i, int j) { return i < j ? Edge{i, j} : Edge{j, i}; }

// Adaptive cost-to-go heuristic: Dijkstra from the goal over the RGG minus the
// known-invalid edges. Optimistic — it does not collision-check (that is the
// forward search's job); the forward search's findings shrink this graph next
// batch (Strub & Gammell 2020).
std::vector<double> reverse_search(SamplingSpace<Point>& space, const std::vector<Point>& points,
                                   const std::vector<std::vector<int>>& nbr,
                                   const std::set<Edge>& invalid_edges, int goal_idx) {
  int n = static_cast<int>(points.size());
  std::vector<double> h(static_cast<size_t>(n), kInf);
  std::vector<char> settled(static_cast<size_t>(n), 0);
  h[static_cast<size_t>(goal_idx)] = 0.0;
  using Entry = std::pair<double, int>;  // (dist, vertex)
  std::priority_queue<Entry, std::vector<Entry>, std::greater<Entry>> heap;
  heap.emplace(0.0, goal_idx);
  while (!heap.empty()) {
    auto [d_u, u] = heap.top();
    heap.pop();
    if (settled[static_cast<size_t>(u)]) continue;
    settled[static_cast<size_t>(u)] = 1;
    for (int w : nbr[static_cast<size_t>(u)]) {
      if (invalid_edges.count(make_edge(u, w))) continue;
      double nd = d_u + space.distance(points[static_cast<size_t>(u)], points[static_cast<size_t>(w)]);
      if (nd < h[static_cast<size_t>(w)]) {
        h[static_cast<size_t>(w)] = nd;
        heap.emplace(nd, w);
      }
    }
  }
  return h;
}

std::vector<Point> reconstruct(const std::vector<Point>& points, const std::vector<int>& parent,
                               int goal_idx) {
  std::vector<Point> path;
  for (int node = goal_idx; node != -1; node = parent[static_cast<size_t>(node)]) {
    path.push_back(points[static_cast<size_t>(node)]);
  }
  std::reverse(path.begin(), path.end());
  return path;
}
}  // namespace

core::PlanResult<Point> AitStarPlanner::plan(SamplingSpace<Point>& space, const Point& start,
                                             const Point& goal, TraceRecorder* recorder) {
  const int batch_size = params_.get_int("batch_size");
  const int max_batches = params_.get_int("max_batches");
  const double gamma = params_.get_float("gamma");
  std::mt19937 rng(static_cast<unsigned>(params_.get_int("seed")));

  auto t0 = std::chrono::steady_clock::now();
  const int goal_idx = 1;
  std::vector<Point> points{start, goal};  // 0 = start (root), 1 = goal
  std::set<Edge> invalid_edges;            // grows across batches: adaptive feedback
  double c_best = kInf;
  int expanded = 0;
  // Last batch's forward-search result; the final incumbent is read from it.
  std::vector<double> g{0.0, kInf};
  std::vector<int> parent{-1, -1};

  for (int batch = 0; batch < max_batches; ++batch) {
    // --- 1. grow the RGG (informed once a solution exists) ---------------------
    int drawn = 0;
    for (int attempt = 0; attempt < batch_size * 40 && drawn < batch_size; ++attempt) {
      Point q = informed_sample(space, start, goal, c_best, rng);
      if (!space.is_state_valid(q)) continue;
      points.push_back(q);
      ++drawn;
      if (recorder) recorder->sample_drawn(q);
    }

    int n = static_cast<int>(points.size());
    double radius = rgg_radius(gamma, n);
    std::vector<std::vector<int>> nbr = radius_neighbors(space, points, radius);

    // --- 2+3. reverse search over the filtered graph = adaptive heuristic ------
    std::vector<double> h_hat = reverse_search(space, points, nbr, invalid_edges, goal_idx);

    // --- 4. forward A* keyed on g + h_hat, lazily validating each edge ---------
    g.assign(static_cast<size_t>(n), kInf);
    g[0] = 0.0;
    parent.assign(static_cast<size_t>(n), -1);
    std::vector<char> closed(static_cast<size_t>(n), 0);
    using Entry = std::pair<double, int>;  // (key, vertex)
    std::priority_queue<Entry, std::vector<Entry>, std::greater<Entry>> open_heap;
    open_heap.emplace(h_hat[0], 0);
    while (!open_heap.empty()) {
      auto [key, v] = open_heap.top();
      open_heap.pop();
      // Lazy deletion: skip entries superseded by a cheaper g[v] (same trick BIT*
      // uses for its vertex queue).
      if (closed[static_cast<size_t>(v)] ||
          key > g[static_cast<size_t>(v)] + h_hat[static_cast<size_t>(v)] + 1e-9)
        continue;
      closed[static_cast<size_t>(v)] = 1;
      ++expanded;
      for (int x : nbr[static_cast<size_t>(v)]) {
        if (invalid_edges.count(make_edge(v, x))) continue;
        double d = space.distance(points[static_cast<size_t>(v)], points[static_cast<size_t>(x)]);
        if (!space.is_motion_valid(points[static_cast<size_t>(v)], points[static_cast<size_t>(x)])) {
          // Discovered invalid: exclude it from every future batch's reverse
          // search — this is AIT*'s adaptive feedback loop.
          invalid_edges.insert(make_edge(v, x));
          continue;
        }
        double new_g = g[static_cast<size_t>(v)] + d;
        if (new_g < g[static_cast<size_t>(x)]) {
          bool first = g[static_cast<size_t>(x)] == kInf;
          g[static_cast<size_t>(x)] = new_g;
          parent[static_cast<size_t>(x)] = v;
          if (recorder) {
            // Emit candidate_evaluated only for feasible, improving edges (not
            // every relaxed neighbour) so the trace stays renderable by replay.py
            // — matches BIT*'s emit-on-accept scale rather than exploding on the
            // batch-recomputed graph.
            recorder->candidate_evaluated(points[static_cast<size_t>(x)], new_g);
            if (first)
              recorder->edge_added(points[static_cast<size_t>(x)], points[static_cast<size_t>(v)], d);
            else
              recorder->rewire(points[static_cast<size_t>(x)], points[static_cast<size_t>(v)]);
          }
          open_heap.emplace(new_g + h_hat[static_cast<size_t>(x)], x);
          if (x == goal_idx && new_g < c_best) {
            c_best = new_g;
            if (recorder) recorder->path_found(reconstruct(points, parent, goal_idx));
          }
        }
      }
    }
  }

  // --- extract final incumbent -----------------------------------------------
  int n = static_cast<int>(points.size());
  bool success = g[static_cast<size_t>(goal_idx)] < kInf;
  std::vector<Point> path = success ? reconstruct(points, parent, goal_idx) : std::vector<Point>{};
  double cost = success ? path_length(space, path) : 0.0;

  double rt = std::chrono::duration<double>(std::chrono::steady_clock::now() - t0).count();
  emit_finished_batch(recorder, success, cost, expanded, n, n, rt);
  core::PlanResult<Point> result;
  result.success = success;
  result.path = std::move(path);
  result.cost = cost;
  result.stats = {expanded, n, expanded, n};
  return result;
}

}  // namespace navigation::global_planning
